File: laravel/app/Http/Controllers/Api/Tournaments/AwardedPrizeController.php
<?php

namespace App\Http\Controllers\Api\Tournaments;

use App\Http\Controllers\Controller;
use Illuminate\Http\Request;
use Illuminate\Http\JsonResponse;
use App\Models\Tournaments\AwardedPrize;
use App\Models\Tournaments\TournamentPrize;
use App\Models\Players\Player;

class AwardedPrizeController extends Controller
{

    public function index(): JsonResponse
    {
        return response()->json(AwardedPrize::all());
    }

    public function show(AwardedPrize $awardedPrize): JsonResponse
    {
        return response()->json($awardedPrize);
    }

    public function claim(Request $request, AwardedPrize $awardedPrize): JsonResponse
    {
        $awardedPrize->claim();
        $awardedPrize->save();
        return response()->json(null, 204);
    }
}

File: phoenix/lib/cards_project_web/controllers/tournaments/awarded_prize_controller.ex
defmodule CardsProjectWeb.Tournaments.AwardedPrizeController do
  use Phoenix.Controller, formats: [:json]
  import Plug.Conn

  alias CardsProject.Tournaments
  alias CardsProject.Tournaments.AwardedPrize

  def index(conn, _params) do
    awarded_prizes = Tournaments.list_awarded_prizes()
    json(conn, Enum.map(awarded_prizes, &serialize_awarded_prize/1))
  end

  def show(conn, %{"id" => id}) do
    awarded_prize = Tournaments.get_awarded_prize!(id)
    json(conn, serialize_awarded_prize(awarded_prize))
  end

  # POST /api/awarded-prizes/{id}/claim
  def claim(conn, %{"id" => id}) do
    Tournaments.awarded_prize_claim_behavior(id)
    send_resp(conn, :no_content, "")
  end

  defp serialize_awarded_prize(%AwardedPrize{} = record) do
    record
    |> Map.take([:id, :final_placement, :awarded_at, :claimed, :claimed_at, :prize_id, :player_id])
    |> (fn m -> Map.put(Map.delete(m, :awarded_at), :awarded_at, Map.get(m, :awarded_at)) end).()
    |> (fn m -> Map.put(Map.delete(m, :claimed_at), :claimed_at, Map.get(m, :claimed_at)) end).()
  end

end
